### ISSUE TITLE
Unable to set clang-7 CXX compiler in Cmake

### ISSUE THREAD

**Issue by breznak** [htm-community/htm.core#521]
```
- Performing Test compiler_supports_machine_option
-- Performing Test compiler_supports_machine_option - Success
-- CMAKE_CXX_COMPILER_ID= Clang
-- CMAKE_CXX_COMPILER_VERSION=7.0.0
-- CMAKE_CXX_COMPILER   = /usr/bin/clang++-7
-- CMAKE_CXX_STANDARD   = 17
-- PLATFORM             = linux
-- BITNESS              = 64
-- NEEDS_BOOST          = OFF
-- INTERNAL_CXX_FLAGS   = -fvisibility-inlines-hidden;-fdiagnostics-show-option;-m64;-fPIC;-Wno-deprecated-register;-Werror;-Wextra;-Wreturn-type;-Wunused;-Wno-unused-variable;-Wno-unused-parameter;-Wno-missing-field-initializers;-Wall;-pipe;-O3;-mtune=generic
-- INTERNAL_LINKER_FLAGS= -m64;-Wl,-undefined,error;-O3
-- INTERNAL_LINKER_FLAGS_STR= -m64 -Wl,-undefined,error -O3
-- COMMON_COMPILER_DEFINITIONS_STR=-DHAVE_UNISTD_H -DNTA_COMPILER_CLANG
-- COMMON_OS_LIBS       = pthread;dl
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:46 (project):
The CMAKE_CXX_COMPILER:
/usr/bin/clang++
is not a full path to an existing compiler tool.
Tell CMake where to find the compiler by setting either the environment
variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
to the compiler, or to the compiler name if it is in the PATH.
-- Configuring incomplete, errors occurred!
See also "/mnt/store/devel/HTM/htm-community/nupic.cpp/build/ThirdParty/CMakeFiles/CMakeOutput.log".
See also "/mnt/store/devel/HTM/htm-community/nupic.cpp/build/ThirdParty/CMakeFiles/CMakeError.log".
CMake Error at external/bootstrap.cmake:49 (message):
CMake step for Third Party builds failed: 1
Call Stack (most recent call first):
CMakeLists.txt:79 (include)
-- Configuring incomplete, errors occurred!
```

I'm using 
> $ CMAKE_CXX_COMPILER=clang++-7 cmake ../.. -DCMAKE_BUILD_TYPE=Release 

or 

> export CXX=clang++-7

**Comment by breznak** [comment]
ok, the problem was that

> /usr/bin/clang++
is not a full path to an existing compiler tool.

a) (on Ubuntu) I've uninstalled clang-6 which provides package "clang", I now only have clang-7, clang-9 
b) ThirdParty deps have been build with a different compiler, that caused the error. 

Workaround was: manually rm the folder build/ThirdParty and rebuild. 

This issue can probably be closed. 

Can we: 
1) ideally rebuild ThirdParty when compiler changes? Cmake should handle that
2) atleast warn in such case? 

CC @dkeeney

**Comment by dkeeney** [comment]
> ideally rebuild ThirdParty when compiler changes? Cmake should handle that

This is difficult because the cmake for ThirdParty does not really get ran once the ThirdParty targets have been built.  If anything changes in ThirdParty, it requires the ThirdParty folder to be deleted before it will be picked up.  There is probably some form of `make clean` that would do the same thing.

However, I will think about this and if I find a way to fix this without re-downloading everything each time I will fix it.

**Comment by breznak** [comment]
> However, I will think about this and if I find a way to fix this without re-downloading everything each time I will fix it. 

on the other hand, how ThirdParty works now is really convenient (that it does build only once), and this is a minor issue, so maybe just a doc not, maybe not even that and only the memo in this issue would be enough?

**Comment by dkeeney** [comment]
> ... and only the memo in this issue would be enough?

Perhaps, but that is not the only issue.  If any of the modules we download should change on their respective download sites, ThirdParty does not see the change once it has completed a download.  So, to re-fetch all of the third party packages the build/ThirdParty folder must be deleted before doing a build.  

Perhaps the best we can do is add a comment in the README.md that explains this.  How about this wording:
```
Third party components needed by this package are automatically downloaded
and built during the first build.  Once the Third party components have been 
downloaded and built they will not be re-visited again on subsequent builds.  
So to refresh the third party components or rebuild them, delete the folder 
build/ThirdParty and then re-build.
```

**Comment by breznak** [comment]
> If any of the modules we download should change on their respective download sites 

but that should not ever happen? We are responsible for bumping version numbers when needed..and then the cmake would rebuild, no? 

>  README.md that explains this. How about this wording:

this sounds good to me!

**Comment by dkeeney** [comment]
> but that should not ever happen? We are responsible for bumping version numbers when needed..and then the cmake would rebuild, no?

Well, yes it can happen.  YamlCpp and mnist are downloading directly from their repository masters.  If those change our re-builds would not see it unless the ThirdParty folder is deleted.  

I would really like to get real release downloads for those two.  mnist does not seem to have versions and probably will not change anyway.   YamlCpp release 0.6.2 is messed up so cannot be used. The previous release is missing some things we need.  The master did change about 3 months ago, unfortunately nothing important so the bug we have with it is still there.
